### PR TITLE
Update linalg.py

### PR DESCRIPTION
**Pull request by cdfreeman-google**
Adding special casing for 3x3 matrices, which speeds up det calculation by ~8x

### CODE REVIEW

File: jax/_src/numpy/linalg.py [hunk 1]
```diff
@@ -238,11 +238,32 @@ def _cofactor_solve(a, b):
...
+def _det_3x3(a):
+  return (a[..., 0, 0] * a[..., 1, 1] * a[..., 2, 2] +
+          a[..., 0, 1] * a[..., 1, 2] * a[..., 2, 0] +
+          a[..., 0, 2] * a[..., 1, 0] * a[..., 2, 1] -
+          a[..., 0, 2] * a[..., 1, 1] * a[..., 2, 0] -
+          a[..., 0, 0] * a[..., 1, 2] * a[..., 2, 1] -
+          a[..., 0, 1] * a[..., 1, 0] * a[..., 2, 2])
+
+@jit
+def _det(a):
```
Comment (hawkinsp) [inline jax/_src/numpy/linalg.py:250]: I think you can just stick a `@jit` around `det`, no need for another `_det` function.

(Sometimes we use that pattern when we want to handle static arguments specially, but `det` doesn't have any.)

File: jax/_src/numpy/linalg.py [hunk 2]
```diff
@@ -238,11 +238,32 @@ def _cofactor_solve(a, b):
   return partial_det[..., -1], x
 
 
+def _det_3x3(a):
```
Comment (hawkinsp) [inline jax/_src/numpy/linalg.py:241]: Is it worth adding a 2x2 case also?

**Review by hawkinsp — CHANGES_REQUESTED** [review]

File: jax/_src/numpy/linalg.py (hunk 2 above)
Comment (cdfreeman-google) [reply jax/_src/numpy/linalg.py:241]: ooh good call!

File: jax/_src/numpy/linalg.py (hunk 1 above)
Comment (cdfreeman-google) [reply jax/_src/numpy/linalg.py:250]: kk--was about to ask about this actually :D

Comment (hawkinsp) [reply jax/_src/numpy/linalg.py:250]: BTW the jit should probably be innermost (i.e., closest to the `def`)

File: jax/_src/numpy/linalg.py [hunk 3]
```diff
@@ -237,12 +237,35 @@ def _cofactor_solve(a, b):
 
   return partial_det[..., -1], x
 
+def _det_2x2(a):
+  return ( a[..., 0, 0] * a[..., 1, 1] -
```
Comment (hawkinsp) [inline jax/_src/numpy/linalg.py:241]: Remove extra space after `(`

File: jax/_src/numpy/linalg.py (hunk 1 above)
Comment (cdfreeman-google) [reply jax/_src/numpy/linalg.py]: ++

**Review by hawkinsp — COMMENTED** [review]
Please squash your commits.

File: jax/_src/numpy/linalg.py (hunk 1 above)
Comment (cdfreeman-google) [reply jax/_src/numpy/linalg.py]: done, I think!

File: jax/_src/numpy/linalg.py (hunk 2 above)
Comment (cdfreeman-google) [reply jax/_src/numpy/linalg.py:244]: also done!

**Comment by hawkinsp** [comment]
A version of this PR was already merged as https://github.com/google/jax/pull/6543

**Pull request closed without merging**